### PR TITLE
feature/update employee screen

### PR DESCRIPTION
**Pull request by nguyenvinhduyet**
## Change Description 👋

- Feature: update employee
- Fix: fix show-all show-enable show-disable
- Fix: adding page container for all my tasks

## Demo 📺

**[Loom Video Demo](https://www.loom.com/share/c8cab56d77584cb28634e8f35c7c08f7)**

**[Loom Video Demo Update (resolve requests change)](https://www.loom.com/share/883f179dee4640d299480d44993a49ff)**

### CODE REVIEW

**Comment by luantc13** [comment]
We don't need to confirm enable/disable in the updating page.
image

**Comment by luantc13** [comment]
What happens if we click to `Save` button after deleted the employee?

File: src/employees/components/UpdateEmployeeForm.vue [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+          10
+        ),
+        validators.min("Phone number must be more than 9 characters", 10),
+      ],
+    };
+  },
+  watch: {
+    employee() {
+      this.employeeCopy = _.cloneDeep(this.employee);
+      this.employeeCopy.birthDate = this.employeeCopy.birthDate.slice(0, 10);
```
Comment (luantc13) [inline src/employees/components/UpdateEmployeeForm.vue:199]: What is this magic?

File: src/employees/components/UpdateEmployeeForm.vue [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+    confirmChangeStatus(status) {
+      const actionTypeName =
+        status === this.ENABLED_STATUS
+          ? "Disable"
+          : status === this.DISABLED_STATUS
+          ? "Enable"
+          : status
+          ? "Enable"
+          : "Disable";
```
Comment (luantc13) [inline src/employees/components/UpdateEmployeeForm.vue:225]: Pls, don't use this magic anymore, `if else` is the best choice here.

File: src/employees/components/UpdateEmployeeForm.vue [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+      this.actionType = "changeStatus"
+      this.isDialog = true
+      this.dialogTitle = `${actionTypeName} employee!`;
+      this.dialogContent = `Are you sure you want to ${actionTypeName.toLowerCase()} the employee named "${this.employeeCopy.firstName +" "+ this.employeeCopy.lastName}"?`;
```
Comment (luantc13) [inline src/employees/components/UpdateEmployeeForm.vue:229]: `${this.employeeCopy.firstName +" "+ this.employeeCopy.lastName}` is not a good approach.
Let's try with:
`${this.employeeCopy.firstName} ${this.employeeCopy.lastName}`

File: src/employees/components/UpdateEmployeeForm.vue [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    },
+
+    confirmDeleteEmployee(employeeId) {
+      this.selectedEmployeeId = employeeId;
+      this.actionType = "deleteEmployee";
+      this.isDialog = true
+      this.dialogTitle = 'Confirm delete employee!';
+      this.dialogContent = `Are you sure want to delete the employee '${this.employeeCopy.firstName +" "+ this.employeeCopy.lastName}'?`
```
Comment (luantc13) [inline src/employees/components/UpdateEmployeeForm.vue:237]: The same here.

File: src/employees/components/UpdateEmployeeForm.vue [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    },
+
+    changeStatus() {
+      if (this.employeeCopy.status === DISABLE_STATUS) return this.employeeCopy.status = ENABLE_STATUS
+      else return this.employeeCopy.status = DISABLE_STATUS
```
Comment (luantc13) [inline src/employees/components/UpdateEmployeeForm.vue:242]: We should have {} for all if else statements, it's a javascript convention.

Comment (luantc13) [reply src/employees/components/UpdateEmployeeForm.vue:242]:
```suggestion
      if (this.employeeCopy.status === DISABLE_STATUS) {
        return this.employeeCopy.status = ENABLE_STATUS;
      } else {
        return this.employeeCopy.status = DISABLE_STATUS;
      }
```

File: src/employees/store/index.js [hunk 6]
```diff
@@ -10,6 +10,7 @@ export const namespaced = true;
 const state = {
   employees: [],
   totalPages: 0,
+  employee: {},
```
Comment (luantc13) [inline src/employees/store/index.js:13]: The default value should be `undefined`, not an empty object `{}`.

File: src/employees/views/UpdateEmployee.vue [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  created() {
+    this.$store.dispatch("employees/getEmployeeById", {
+      companyId: this.$route.params.companyId,
+      employeeId: this.$route.params.employeeId,
+    });
+  },
+};
+</script>
+
+<style></style>
```
Comment (luantc13) [inline src/employees/views/UpdateEmployee.vue:49]: All files should have an empty line at the end.

File: src/http/employees.js [hunk 8]
```diff
@@ -18,5 +18,16 @@ const updateStatusEmployee = (companyId, employeeId, payload) =>
...
+};
 
-export { create, getEmployees, updateStatusEmployee, deleteEmployee }
\ No newline at end of file
+const getEmployeeById = ({ companyId, employeeId }) => {
+  return httpClient.get(`${COMPANY}/${companyId}${EMPLOYEE}/${employeeId}`);
+};
+
+
+export { create, getEmployees, updateStatusEmployee, deleteEmployee, updateEmployee, getEmployeeById }
```
Comment (luantc13) [inline src/http/employees.js:33]: All files should have an empty line at the end.

File: src/products/components/UpdateProductForm.vue [hunk 9]
```diff
@@ -148,6 +148,7 @@ export default {
   watch: {
     product() {
       this.productCopy = _.cloneDeep(this.product);
+      console.log(this.productCopy)
```
Comment (luantc13) [inline src/products/components/UpdateProductForm.vue:151]: We should remove all `console.log` before commiting.

**Review by luantc13 — CHANGES_REQUESTED** [review]
Great work!
Having some comments, happy to merge when they are resolved.

**Comment by nguyenvinhduyet** [comment]
> What happens if we click to `Save` button after deleted the employee?
- It will display an error message: that this information does not exist
image

-> Fix: after pressing the delete button I will be redirected to the employee list page

**Review by luantc13 — APPROVED** [review]
Nice work!

File: src/employees/components/UpdateEmployeeForm.vue [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+            >SAVE</v-btn
+          >
+        </v-col>
+
+        <v-col cols="12" md="2">
+          <v-btn outlined color="#F44336" @click="confirmDeleteEmployee(employeeCopy.id)"
+            >DELETE</v-btn
+          >
+        </v-col>
+      </v-row>
```
Comment (truonghungit) [inline src/employees/components/UpdateEmployeeForm.vue:132]: We should align the delete button to the right 
image


simple yet best solution

```suggestion
      <div class="d-flex justify-space-between">
        <div>
          <v-btn outlined color="#72418b" @click="cancelUpdateEmployee()" class="mr-5"
            >CANCEL</v-btn
          >

          <v-btn color="#72418b" dark type="submit" @click="validate()" class="mr-5"
            >SAVE</v-btn
          >
        </div>

        <v-btn outlined color="#F44336" @click="confirmDeleteEmployee(employeeCopy.id)"
          >DELETE</v-btn
        >
      </div>
```

Comment (truonghungit) [reply src/employees/components/UpdateEmployeeForm.vue:132]: I believe the component `CreateEmployeeForm` can be reused for updates, just rename it to `EmployeeForm` and update small things to binding data to UI (in edit mode), therefore we don't need the `UpdateEmployeeForm.vue`

in the CreateEmployee page, we will have
`<EmployeeForm @submit="onCreateEmployee" />`

in the UpdateEmployee page we will have
`<EmployeeForm :employee="employee" @submit="onUpdateEmployee" @delete-employee="onDeleteEmployee"/>`

**Review by truonghungit — CHANGES_REQUESTED** [review]

**Comment by nguyenvinhduyet** [comment]
video loom demo update: https://www.loom.com/share/22b13229dd734e5e847eb300d31d3e4d

**Comment by nguyenvinhduyet** [comment]
Video demo update (fix: direct page "employee not found"): https://www.loom.com/share/873afc7534ab42978daeffd1fb62a275

**Review by truonghungit — APPROVED** [review]

**Pull request merged**